gapir/cc/memory_manager.cpp

```diff
@@ -78,48 +80,15 @@ MemoryManager::MemoryManager(const std::vector<uint32_t>& sizeList)
   }
 
   GAPID_DEBUG("Base address: %p", mMemory.get());
-  setReplayDataSize(0, 0);
   setVolatileMemory(mSize);
 }
 
-bool MemoryManager::setReplayDataSize(uint32_t constantMemorySize,
-                                      uint32_t opcodeMemorySize) {
-  GAPID_DEBUG("MemoryManager::setReplayDataSize(%d, %d)", constantMemorySize,
-              opcodeMemorySize);
-  if (opcodeMemorySize > mSize) {
-    GAPID_ERROR("Opcode memory size: %d larger than total memory size: %d",
-                opcodeMemorySize, mSize);
-    return false;
-  }
-  mOpcodeMemory = {align(mMemory.get() + mSize - opcodeMemorySize),
-                   opcodeMemorySize};
-  GAPID_DEBUG("Opcode range: [%p,%p]", mOpcodeMemory.base,
-              mOpcodeMemory.base + mOpcodeMemory.size - 1);
-
-  if (constantMemorySize > mSize - mOpcodeMemory.size) {
-    GAPID_ERROR(
-        "Constant memory size: %d larger than available memory size: %d",
-        constantMemorySize, mSize - mOpcodeMemory.size);
-    return false;
-  }
-  mConstantMemory = {align(mOpcodeMemory.base - constantMemorySize),
-                     constantMemorySize};
-  GAPID_DEBUG("Constant range: [%p,%p]", mConstantMemory.base,
-              mConstantMemory.base + mConstantMemory.size - 1);
-
-  mReplayData = {mConstantMemory.base,
-                 mConstantMemory.size + mOpcodeMemory.size};
-  GAPID_DEBUG("Replay range: [%p,%p]", mReplayData.base,
-              mReplayData.base + mReplayData.size - 1);
-  return true;
-}
-
 bool MemoryManager::setVolatileMemory(uint32_t size) {
-  if (size > mSize - mReplayData.size) {
+  if (size > mSize) {
     return false;
   }
 
-  mVolatileMemory = {align(mReplayData.base - size), size};
+  mVolatileMemory = {align(&mMemory[0] + mSize - size), size};
```

Comment:
Any reason you're allocating from the top instead of the bottom?
Depending on the alignment of `mMemory`, it looks like this might be possible to underflow the buffer as the alignment happens after the bounds check.

Reply:
This was for consistency, but I think for future changes volatile memory should just be at the bottom of the memory manager, (and the cache can live at the top). So I will make that change now, with only 2 memory regions this is a lot easier.